stock_resupply_order/models/stock_resupply_order.py

```diff
@@ -0,0 +1,200 @@
+        "Company",
+        required=True,
+        index=True,
+        default=lambda self: self.env.company,
+    )
+
+    stock_resupply_order_line_ids = fields.One2many(
+        "stock.resupply.order.line",
+        inverse_name="stock_resupply_order_id",
+        help="All product quantities desired at the given location",
+        string="Desired products quantities",
+    )
+
+    procurement_group_id = fields.Many2one(
+        "procurement.group", string="Generated procurement"
+    )
+
+    state = fields.Selection(
+        [
+            ("draft", "Draft"),
+            ("done", "Done"),
+        ],
+        "Status",
+        help=(
+            "Resupply order state: \n"
+            "* draft: The order has not been confirmed yet.\n"
+            "* done: The procurements have been created. \n"
+        ),
+        copy=False,
+        readonly=True,
+        tracking=True,
+        store=True,
+        default="draft",
+    )
+
+    picking_ids = fields.One2many(
+        "stock.picking",
+        compute="_compute_picking_ids",
+        string="Generated stock pickings",
+        readonly=True,
+    )
+
+    picking_count = fields.Integer(
+        string="Picking count", compute="_compute_picking_ids", readonly=True
+    )
+
+    @api.depends("procurement_group_id.stock_move_ids.picking_id")
+    def _compute_picking_ids(self):
+        for record in self:
+            record.picking_ids = record.procurement_group_id.stock_move_ids.picking_id
+            record.picking_count = len(record.picking_ids)
+
+    def action_run(self):
+        if self.procurement_group_id:
+            return self.procurement_group_id
+
+        self.procurement_group_id = self.env["procurement.group"].create(
+            self._get_procurement_group_parameters()
+        )
+
+        quant_groups = self._get_existing_quants()
+
+        procurements = []
+
+        for line in self.stock_resupply_order_line_ids:
+            # service products have no tracking/lot and cannot be run in procurement
+            if line.product_id.product_tmpl_id.type == "service":
+                continue
+
+            available_quantity = self._get_available_quantity_for_product(
+                quant_groups, line
+            )
+
+            if available_quantity < line.quantity:
+                procurements.append(
+                    self.env["procurement.group"].Procurement(
+                        product_id=line.product_id,
+                        product_qty=line.quantity - available_quantity,
+                        product_uom=line.product_id.product_tmpl_id.uom_id,
+                        location_id=self.location_id,
+                        name=f"Resupply Order to {self.location_id.name}",
+                        origin=f"Resupply Order to {self.location_id.name}",
+                        company_id=self.company_id,
+                        values=self._get_procurement_values(),
+                    )
+                )
```

Comment:
This could be splitted for latter easiest overrides

```suggestion
            proc = self._prepare_resupply_procurement(line, quant_groups)
            if proc:
                procurements.append(proc)
```


```
def _prepare_resupply_procurement(self, resupply_line, quant_groups)
        # service products have no tracking/lot and cannot be run in procurement
        if resupply_line.product_id.product_tmpl_id.type == "service":
            return

        available_quantity = self._get_available_quantity_for_product(
            quant_groups, resupply_line
        )

        if available_quantity >= resupply_line.quantity:
            return

        name = origin = f"Resupply Order to {self.location_id.name}"
        return self.env["procurement.group"].Procurement(
              name=name,
              origin=origin,
              product_id=resupply_line.product_id,
              product_qty=resupply_line.quantity - available_quantity,
              product_uom=resupply_line.product_id.product_tmpl_id.uom_id,
              location_id=self.location_id,
              company_id=self.company_id,
              values={"group_id": self.procurement_group_id},
          )
```